File: scripts/inject-exif.js
#!/usr/bin/env node
/**
 * inject-exif.js — Injection XMP via ExifTool (sans ré-encodage)
 * Source de vérité : Supabase (photos_sous_marine + photos_paysage)
 * MISC_FILES pour images presse/médias/divers hors galeries
 * Usage: node scripts/inject-exif.js
 */

import { execSync }                                                        from 'child_process';
import { existsSync, readFileSync, writeFileSync, unlinkSync, statSync }   from 'fs';
import { join, dirname }                                                    from 'path';
import { fileURLToPath }                                                    from 'url';
import { tmpdir }                                                           from 'os';
import { createClient }                                                     from '@supabase/supabase-js';

const NEW_ONLY    = process.argv.includes('--new-only');
const CACHE_FILE  = join(dirname(fileURLToPath(import.meta.url)), '..', '.exif-cache.json');

const __filename = fileURLToPath(import.meta.url);
const __dirname  = dirname(__filename);
const path = { join };

const ROOT        = path.join(__dirname, '..');
const IMAGES_ROOT = path.join(ROOT, 'public', 'images');

const CREATOR       = 'Karim Saari';
const COPYRIGHT     = '(c) Karim Saari - Dark Massilia - karimsaari.com';
const WEB_STATEMENT = 'https://karimsaari.com';
const USAGE_TERMS   = 'All rights reserved - Contact: email@karimsaari.com';

const GEO_MARSEILLE = { lat: '43.2965', lon: '5.3698', city: 'Marseille', country: 'France', state: 'Provence-Alpes-Cote d\'Azur' };

// ── Keywords ──────────────────────────────────────────────────────────────────
const KW_BASE     = ['karim saari', 'dark massilia', 'marseille', 'méditerranée'];
const KW_SOUS_MER = [...KW_BASE, 'photographe sous-marin', 'calanques', 'apnée', 'photographie sous-marine', 'calanques de marseille'];
const KW_DEPOLL   = [...KW_SOUS_MER, 'dépollution marine', 'pollution plastique', 'projet sentinelle', 'dépollution sous-marine'];
const KW_BIOD     = [...KW_SOUS_MER, 'biodiversité marine', 'faune méditerranéenne', 'vie marine'];
const KW_CARACT   = [...KW_SOUS_MER, 'caractérisation déchets', 'projet sentinelle'];
const KW_PAYSAGE  = [...KW_BASE, 'photographe paysage', 'photographie de paysage', 'calanques'];
const KW_MER      = [...KW_PAYSAGE, 'calanques de marseille', 'littoral méditerranéen', 'paysage marin'];
const KW_TERRE    = [...KW_PAYSAGE, 'provence', 'lavande', 'valensole', 'photographie nature'];
const KW_HORIZONS = [...KW_BASE, 'photographe paysage', 'photographie de paysage', 'paysage', 'voyage', 'nature'];
// MISC_FILES only
const KW_PRESSE   = [...KW_BASE, 'calanques', 'photographe sous-marin', 'dépollution marine', 'projet sentinelle'];
const KW_ARTE     = [...KW_BASE, 'arte', 'documentaire', 'photographe sous-marin'];
const KW_RUGULO   = [...KW_BASE, 'calanques', 'marseilleveyre', 'rugulopteryx okamurae', 'algue invasive', 'biodiversité marine'];
const KW_DIVERS   = [...KW_BASE, 'calanques', 'photographe sous-marin'];
const KW_MISSIONS = [...KW_DEPOLL];

// ── .env loader ───────────────────────────────────────────────────────────────
function loadEnv() {
  const envPath = path.join(ROOT, '.env');
  if (!existsSync(envPath)) return {};
  return Object.fromEntries(
    readFileSync(envPath, 'utf-8')
      .split('\n')
      .filter(l => l.trim() && !l.startsWith('#'))
      .map(l => { const i = l.indexOf('='); return [l.slice(0, i).trim(), l.slice(i + 1).trim()]; })
      .filter(([k]) => k)
  );
}

const EXIFTOOL_PATH = 'C:\\Users\\ksaari\\AppData\\Local\\Programs\\ExifTool\\ExifTool.exe';

// ── Cache (mtime) pour --new-only ─────────────────────────────────────────────
function loadCache() {
  try { return JSON.parse(readFileSync(CACHE_FILE, 'utf-8')); } catch { return {}; }
}
function saveCache(cache) {
  writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2), 'utf-8');
}
function isAlreadyProcessed(cache, filePath) {
  if (!NEW_ONLY) return false;
  const mtime = statSync(filePath).mtimeMs;
  return cache[filePath] === mtime;
}
function markProcessed(cache, filePath) {
  cache[filePath] = statSync(filePath).mtimeMs;
}

// ── ExifTool via argfile cp1252 (encodage Windows natif) ─────────────────────
function runExiftool(filePath, title, description, keywords, geo) {
  // Étape 1 : purge de tous les Subject existants via appel direct
  execSync(
    `"${EXIFTOOL_PATH}" "-XMP:Subject=" -overwrite_original "${filePath}"`,
    { stdio: 'pipe' }
  );

  // Étape 2 : écriture via argfile encodé en latin1 (cp1252 Windows)
  // Tous les caractères français (é â ê î ô û à ç — etc.) sont dans le range cp1252
  // ExifTool lit le fichier avec l'encodage système Windows → conversion correcte en UTF-8 XMP
  const args = [
    `-XMP:Title=${title}`,
    `-XMP:Description=${description}`,
    `-XMP:Creator=${CREATOR}`,
    `-XMP:Rights=${COPYRIGHT}`,
    `-XMP:WebStatement=${WEB_STATEMENT}`,
    `-XMP:Marked=True`,
    `-XMP:UsageTerms=${USAGE_TERMS}`,
    `-EXIF:Artist=${CREATOR}`,
    `-EXIF:Copyright=${COPYRIGHT}`,
    // Subject déjà purgé en étape 1 → on ajoute proprement
    ...keywords.map(kw => `-XMP:Subject+=${kw}`),
  ];

  if (geo) {
    args.push(
      `-XMP-iptcExt:LocationShownCity=${geo.city}`,
      `-XMP-iptcExt:LocationShownCountryName=${geo.country}`,
      `-IPTC:City=${geo.city}`,
      `-IPTC:Country-PrimaryLocationName=${geo.country}`,
      `-IPTC:Province-State=${geo.state}`,
      `-GPS:GPSLatitude=${geo.lat}`,
      `-GPS:GPSLatitudeRef=${parseFloat(geo.lat) >= 0 ? 'N' : 'S'}`,
      `-GPS:GPSLongitude=${Math.abs(parseFloat(geo.lon))}`,
      `-GPS:GPSLongitudeRef=${parseFloat(geo.lon) >= 0 ? 'E' : 'W'}`,
    );
  }

  args.push('-overwrite_original', filePath);

  const tmpFile = join(tmpdir(), `exiftool-args-${Date.now()}.txt`);
  // UTF-8 sans BOM — Windows en mode UTF-8 system-wide → ExifTool lit UTF-8 par défaut
  // Le BOM cassait le premier argument de l'argfile (était préfixé \uFEFF)
  writeFileSync(tmpFile, args.join('\n'), 'utf-8');

  try {
    const result = execSync(
      `"${EXIFTOOL_PATH}" -charset UTF8 -charset IPTC=UTF8 -@ "${tmpFile}"`,
      { stdio: 'pipe', encoding: 'utf8' }
    );
    if (!result.includes('1 image files updated')) {
      throw new Error(result.trim() || 'ExifTool: aucune image mise à jour');
    }
  } finally {
    unlinkSync(tmpFile);
  }
}

// ── Helpers DB ────────────────────────────────────────────────────────────────
function countryFromLieu(lieu) {
  if (!lieu) return 'France';
  const l = lieu.toLowerCase();
  if (l.includes('madère') || l.includes('madeira') || l.includes('portugal')) return 'Portugal';
  if (l.includes('maroc') || l.includes('morocco'))  return 'Maroc';
  if (l.includes('espagne') || l.includes('spain'))  return 'Espagne';
  if (l.includes('italie') || l.includes('italy'))   return 'Italie';
  return 'France';
}

function geoFromRow(row) {
  if (row.lat != null && row.lng != null) {
    const parts = (row.lieu || '').split(',').map(s => s.trim());
    return {
      lat:     String(row.lat),
      lon:     String(row.lng),
      city:    parts[0] || 'Marseille',
      country: countryFromLieu(row.lieu),
      state:   '',
    };
  }
  return GEO_MARSEILLE;
}

function keywordsForSousMarine(type) {
  if (!type) return KW_SOUS_MER;
  const t = type.toLowerCase();
  if (t.includes('d') && t.includes('pollution')) return KW_DEPOLL;
  if (t.includes('biodiv'))                        return KW_BIOD;
  if (t.includes('caract'))                        return KW_CARACT;
  return KW_SOUS_MER;
}

function keywordsForPaysage(type) {
  if (!type) return KW_PAYSAGE;
  const t = type.toLowerCase();
  if (t === 'mer')      return KW_MER;
  if (t === 'terre')    return KW_TERRE;
  if (t === 'horizons') return KW_HORIZONS;
  return KW_PAYSAGE;
}

// ── Traitement d'une galerie depuis Supabase ──────────────────────────────────
async function processGallery(sb, tableName, keywordsFn, cache) {
  const { data, error } = await sb.from(tableName).select('src,title,alt,lieu,lat,lng,categorie');
  if (error) { console.error(`❌  ${tableName} : ${error.message}`); return; }

  console.log(`\n📁 ${tableName}  (${data.length} photos)`);
  let ok = 0, skip = 0, cached = 0, err = 0;

  for (const row of data) {
    if (!row.src) continue;
    // Ignorer les variantes srcset (traitées séparément si besoin)
    if (/_(?:400w|800w|1200w)\.webp$/.test(row.src)) continue;

    const filePath = path.join(ROOT, 'public', row.src).replace(/\//g, '\\');
    if (!existsSync(filePath)) {
      console.log(`  ⚠  Introuvable : ${row.src}`);
      skip++;
      continue;
    }

    if (isAlreadyProcessed(cache, filePath)) {
      cached++;
      continue;
    }

    const title       = row.title?.trim() || row.src.split('/').pop().replace('.webp', '');
    const description = row.alt?.trim()   || title;
    const keywords    = keywordsFn(row.categorie);
    const geo         = geoFromRow(row);

    try {
      runExiftool(filePath, title, description, keywords, geo);
      const filename = row.src.split('/').pop();
      markProcessed(cache, filePath);
      console.log(`  ✓  ${filename}`);
      ok++;

      // Variante 800w dans le sous-dossier 800w/ du même répertoire
      const dir800w    = path.join(ROOT, 'public', row.src, '..', '800w').replace(/\//g, '\\');
      const file800w   = path.join(dir800w, filename).replace(/\//g, '\\');
      if (existsSync(file800w)) {
        runExiftool(file800w, title, description, keywords, geo);
        markProcessed(cache, file800w);
        console.log(`  ✓  800w/${filename}`);
      }
    } catch (e) {
      console.log(`  ✗  ${row.src.split('/').pop()} → ${e.message}`);
      err++;
    }
  }
  const cachedMsg = cached > 0 ? `, ${cached} déjà traités (cache)` : '';
  console.log(`  → ${ok} OK, ${skip} introuvables, ${err} erreurs${cachedMsg}`);
}

// ── MISC_FILES — images presse/médias/divers (hors galeries DB) ───────────────
// geo: null  → GPS original préservé (ne pas écrire)
// geo: omis  → fallback Marseille
const MISC_FILES = [
  { file: 'karim-saari-interview-presse-tiredearth-photographe-sous-marin.webp',       title: 'Interview Tired Earth — Karim Saari photographe sous-marin',                    description: 'Interview presse Tired Earth — Karim Saari apnéiste et photographe sous-marin Marseille (Dark Massilia)', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-marcelle-media-depollution-mer-apnee.webp',            title: 'Marcelle Média — Dépolluer la mer en apnée',                                     description: 'Reportage Marcelle Média sur la dépollution marine en apnée par Karim Saari — Projet Sentinelle Marseille', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-france-bleu-rorqual-cotes-marseille.webp',             title: 'France Bleu — Rorqual au large de Marseille',                                    description: 'France Bleu Provence — Rorqual rarissime aperçu près des côtes de Marseille, témoignage Karim Saari (Dark Massilia)', keywords: [...KW_PRESSE, 'rorqual', 'grand cétacé'] },
  { file: 'karim-saari-marseille-actu-depollution-fonds-marins.webp',                   title: 'Actu.fr — Dépollution fonds marins Marseille',                                   description: 'Reportage Actu.fr sur la dépollution des fonds marins de Marseille par Karim Saari et Team Oxygen — Calanques', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-fondation-mer-projet-sentinelle-calanques.webp',       title: 'Fondation de la Mer — Projet Sentinelle Calanques',                              description: 'La Fondation de la Mer soutient le Projet Sentinelle de dépollution sous-marine dans les Calanques de Marseille', keywords: [...KW_PRESSE, 'fondation de la mer'] },
  { file: 'karim-saari-marseille-la-provence-operation-sentinelle-apnee.webp',          title: 'La Provence — Opération Sentinelle Marseille',                                   description: 'La Provence couvre l\'Opération Sentinelle de dépollution en apnée dans les Calanques de Marseille — Karim Saari', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-france-bleu-goudes-dechets-apneistes.webp',            title: 'France Bleu — 328 kg déchets aux Goudes',                                       description: 'France Bleu Provence — 328 kg de déchets récoltés aux Goudes par les apnéistes de Team Oxygen Marseille', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-ville-reconnaissance-officielle-dark-massilia.webp',   title: 'Ville de Marseille — Reconnaissance officielle Dark Massilia',                  description: 'La Ville de Marseille reconnaît officiellement l\'engagement de Dark Massilia pour la dépollution marine des Calanques', keywords: KW_PRESSE },
  { file: 'karim-saari-marseille-made-in-marseille-provence-tourisme.webp',             title: 'Made in Marseille — Provence tourisme France',                                   description: 'Made in Marseille — Provence classée top destination touristique France, Calanques et Méditerranée', keywords: [...KW_PRESSE, 'tourisme marseille', 'provence'] },
  { file: 'karim-saari-marseille-echappees-belles-reportage-television.webp',           title: 'Échappées Belles — Karim Saari France Télévisions',                             description: 'Reportage Échappées Belles avec Karim Saari (Dark Massilia) au Vallon des Auffes — Marseille, France Télévisions', keywords: [...KW_PRESSE, 'échappées belles', 'france télévisions', 'vallon des auffes'] },
  { file: 'karim-saari-arte-regard-documentaire-calanques-marseille.webp',              title: 'ARTE Regards — Karim Saari Calanques Marseille',                                description: 'Documentaire ARTE Regards — Karim Saari et Team Oxygen contre la pollution plastique dans les Calanques de Marseille', keywords: KW_ARTE },
  { file: 'karim-saari-photo-profil-arte-regard-marseille.webp',                        title: 'Portrait Karim Saari — Documentaire ARTE',                                      description: 'Portrait de Karim Saari photographe sous-marin pour le documentaire ARTE Regards — Marseille, Dark Massilia', keywords: KW_ARTE },
  { file: 'karim-saari-photo-profil-arte-regard-marseille_300w.webp',                   title: 'Portrait Karim Saari — ARTE (300w)',                                             description: 'Portrait de Karim Saari photographe sous-marin pour le documentaire ARTE Regards — Marseille, Dark Massilia', keywords: KW_ARTE },
  { file: 'arte-meduses-souveraines-oceans-documentaire-marseille-2.webp',              title: 'Méduses — Les souveraines des océans (ARTE) 2',                                 description: 'Documentaire ARTE — Méduses les souveraines des océans, contribution Karim Saari Dark Massilia Marseille', keywords: [...KW_ARTE, 'méduses', 'faune marine'] },
  { file: 'arte-meduses-souveraines-oceans-documentaire-marseille-3.webp',              title: 'Méduses — Les souveraines des océans (ARTE) 3',                                 description: 'Documentaire ARTE — Méduses les souveraines des océans, apnée et photographie sous-marine Méditerranée', keywords: [...KW_ARTE, 'méduses', 'faune marine'] },
  { file: 'arte-meduses-souveraines-oceans-documentaire-marseille-4.webp',              title: 'Méduses — Les souveraines des océans (ARTE) 4',                                 description: 'Documentaire ARTE — Méduses les souveraines des océans, Méditerranée et Calanques de Marseille', keywords: [...KW_ARTE, 'méduses', 'faune marine'] },
  { file: 'arte-meduses-souveraines-oceans-documentaire-marseille-5.webp',              title: 'Méduses — Les souveraines des océans (ARTE) 5',                                 description: 'Documentaire ARTE — Méduses les souveraines des océans, photographe sous-marin Karim Saari', keywords: [...KW_ARTE, 'méduses', 'faune marine'] },
  { file: 'marseille-marseilleveyre-avant-rugulopteryx-fonds-marins.webp',              title: 'Marseilleveyre avant Rugulopteryx okamurae',                                    description: 'Fonds marins de Marseilleveyre avant l\'invasion de Rugulopteryx okamurae — Calanques de Marseille', keywords: KW_RUGULO },
  { file: 'marseille-marseilleveyre-avant-rugulopteryx-biodiversite-2.webp',            title: 'Marseilleveyre — biodiversité avant algue invasive',                             description: 'Biodiversité marine de Marseilleveyre avant l\'invasion de Rugulopteryx okamurae — fonds rocheux naturels', keywords: KW_RUGULO },
  { file: 'marseille-marseilleveyre-apres-rugulopteryx-algue-invasive.webp',            title: 'Marseilleveyre après Rugulopteryx okamurae',                                    description: 'Fonds marins de Marseilleveyre après l\'invasion de Rugulopteryx okamurae — algue invasive recouvrant tout', keywords: [...KW_RUGULO, 'espèce invasive', 'alerte environnementale'] },
  { file: 'marseille-marseilleveyre-apres-rugulopteryx-tapis-algues-2.webp',            title: 'Marseilleveyre — tapis algues invasives Rugulopteryx',                          description: 'Tapis d\'algues invasives Rugulopteryx okamurae sur les fonds de Marseilleveyre — Calanques de Marseille', keywords: [...KW_RUGULO, 'espèce invasive', 'alerte environnementale'] },
  { file: 'karim-saari-photographe-sous-marin-marseille-dark-massilia.webp',            title: 'Karim Saari — Photographe sous-marin Marseille',                                description: 'Portrait de Karim Saari, photographe sous-marin et activiste environnemental à Marseille — Dark Massilia', keywords: KW_DIVERS },
  { file: 'karim-saari-photographe-sous-marin-marseille-dark-massilia_480w.webp',       title: 'Karim Saari — Photographe sous-marin Marseille (480w)',                         description: 'Portrait de Karim Saari, photographe sous-marin et activiste environnemental à Marseille — Dark Massilia', keywords: KW_DIVERS },
  { file: 'karim-saari-marseille-dark-massilia-9ans-celebration-projet-sentinelle.webp', title: '9 ans de Dark Massilia — Projet Sentinelle Marseille',                         description: '9 années d\'engagement de Dark Massilia pour la dépollution marine et la sensibilisation environnementale à Marseille', keywords: KW_DIVERS },
  { file: 'karim-saari-marseille-130000-sentinelles-calanques-depollution.webp',        title: '130 000 Sentinelles — Marseille Calanques',                                     description: 'Mobilisation de 130 000 citoyens sentinelles pour la protection des Calanques de Marseille — Dark Massilia', keywords: KW_MISSIONS },
];

// ── Main ──────────────────────────────────────────────────────────────────────
try {
  execSync(`"${EXIFTOOL_PATH}" -ver`, { stdio: 'pipe' });
} catch {
  console.error(`❌  ExifTool introuvable à : ${EXIFTOOL_PATH}`);
  process.exit(1);
}

const env = loadEnv();
if (!env.VITE_SUPABASE_URL || !env.VITE_SUPABASE_ANON_KEY) {
  console.error('❌  VITE_SUPABASE_URL / VITE_SUPABASE_ANON_KEY absents dans .env');
  process.exit(1);
}

const sb = createClient(env.VITE_SUPABASE_URL, env.VITE_SUPABASE_ANON_KEY);
console.log(`🔧  Injection XMP — source : Supabase${NEW_ONLY ? '  [mode --new-only]' : ''}\n`);

const cache = loadCache();

// Galeries depuis la DB (source de vérité)
await processGallery(sb, 'photos_sous_marine', keywordsForSousMarine, cache);
await processGallery(sb, 'photos_paysage',     keywordsForPaysage,     cache);

// Images presse / médias / divers (hors galeries)
console.log('\n📁 MISC — presse / médias / divers');
let okMisc = 0, cachedMisc = 0, errMisc = 0;
for (const { file, title, description, keywords, geo } of MISC_FILES) {
  const filePath = path.join(IMAGES_ROOT, file).replace(/\//g, '\\');
  if (!existsSync(filePath)) { console.log(`  ⚠  Introuvable : ${file}`); continue; }
  if (isAlreadyProcessed(cache, filePath)) { cachedMisc++; continue; }
  const resolvedGeo = geo === null ? null : (geo ?? GEO_MARSEILLE);
  try {
    runExiftool(filePath, title, description, keywords, resolvedGeo);
    markProcessed(cache, filePath);
    console.log(`  ✓  ${file}`);
    okMisc++;
  } catch (e) {
    console.log(`  ✗  ${file} → ${e.message}`);
    errMisc++;
  }
}
const cachedMiscMsg = cachedMisc > 0 ? `, ${cachedMisc} déjà traités (cache)` : '';
console.log(`  → ${okMisc} OK, ${errMisc} erreurs${cachedMiscMsg}`);

saveCache(cache);
console.log('\n✅  Terminé — relance npm run build:full + FTP');

File: scripts/inject-exif.py
#!/usr/bin/env python3
"""
inject-exif.py — Injection XMP dans les WebP SANS ré-encodage (manipulation RIFF binaire)
Usage: python scripts/inject-exif.py
"""

import struct
from pathlib import Path

# ── Config ──────────────────────────────────────────────────────────────────
BASE_DIR      = Path(__file__).parent.parent / "public"
PORTFOLIO_DIR = BASE_DIR / "images/portfolio/Mer"
MISSIONS_DIR  = BASE_DIR / "images"

CREATOR   = "Karim Saari"
COPYRIGHT = "© Karim Saari — Dark Massilia — Photographe sous-marin Marseille"

KEYWORDS_PORTFOLIO = [
    "photographe sous-marin", "marseille", "calanques", "apnée",
    "méditerranée", "dark massilia", "karim saari",
    "photographie sous-marine", "calanques de marseille",
]
KEYWORDS_MISSIONS = [
    "photographe sous-marin", "marseille", "calanques", "apnée",
    "méditerranée", "dark massilia", "karim saari",
    "dépollution marine", "pollution plastique", "projet sentinelle",
]

UNDERWATER_HINTS = [
    "fonds-marins", "nage", "grotte", "poulpe", "spirographe",
    "musée", "diving", "apneiste", "shooting", "soupe", "poseidon",
    "angel", "octopus", "paysage-sous-marin", "teamoxygen-freediving",
    "mer-de-plastique", "moyades", "mer-goudes", "freediving",
    "sous-marin", "subaquatique", "frioul",
]

# ── XMP builder ─────────────────────────────────────────────────────────────
def make_xmp(title, description, creator, copyright_str, keywords):
    kw_items = "".join(f"      <rdf:li>{kw}</rdf:li>\n" for kw in keywords)
    return f'''<?xpacket begin="\ufeff" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description rdf:about=""
      xmlns:dc="http://purl.org/dc/elements/1.1/"
      xmlns:xmpRights="http://ns.adobe.com/xap/1.0/rights/">
      <dc:title><rdf:Alt><rdf:li xml:lang="x-default">{title}</rdf:li></rdf:Alt></dc:title>
      <dc:description><rdf:Alt><rdf:li xml:lang="x-default">{description}</rdf:li></rdf:Alt></dc:description>
      <dc:creator><rdf:Seq><rdf:li>{creator}</rdf:li></rdf:Seq></dc:creator>
      <dc:rights><rdf:Alt><rdf:li xml:lang="x-default">{copyright_str}</rdf:li></rdf:Alt></dc:rights>
      <dc:subject>
        <rdf:Bag>
{kw_items}        </rdf:Bag>
      </dc:subject>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>'''

# ── WebP RIFF helpers ────────────────────────────────────────────────────────
def parse_chunks(data):
    """Retourne [(id_bytes, data_bytes), ...] à partir de l'offset 12 (après RIFF+size+WEBP)."""
    chunks, offset = [], 12
    while offset + 8 <= len(data):
        cid   = data[offset:offset+4]
        csize = struct.unpack_from('<I', data, offset+4)[0]
        cdata = data[offset+8:offset+8+csize]
        chunks.append((cid, cdata))
        offset += 8 + csize + (csize & 1)
    return chunks

def build_webp(chunks):
    payload = b'WEBP'
    for cid, cdata in chunks:
        payload += cid + struct.pack('<I', len(cdata)) + cdata
        if len(cdata) & 1:
            payload += b'\x00'
    return b'RIFF' + struct.pack('<I', len(payload)) + payload

def vp8_dims(d):
    if d[3:6] == b'\x9d\x01\x2a':
        return (struct.unpack_from('<H', d, 6)[0] & 0x3fff) + 1, \
               (struct.unpack_from('<H', d, 8)[0] & 0x3fff) + 1
    return None, None

def vp8l_dims(d):
    if d[0:1] == b'\x2f':
        bits = struct.unpack_from('<I', d, 1)[0]
        return (bits & 0x3fff) + 1, ((bits >> 14) & 0x3fff) + 1
    return None, None

def inject_xmp(filepath, xmp_str):
    """Injecte XMP dans un WebP sans ré-encoder les pixels."""
    data = Path(filepath).read_bytes()
    if data[:4] != b'RIFF' or data[8:12] != b'WEBP':
        raise ValueError("Pas un fichier WebP valide")

    chunks   = parse_chunks(data)
    chunk_ids = [c[0] for c in chunks]
    xmp_bytes = xmp_str.encode('utf-8')

    if b'VP8X' in chunk_ids:
        # Format étendu — mettre à jour le flag XMP + remplacer le chunk XMP
        new_chunks = []
        for cid, cdata in chunks:
            if cid == b'VP8X':
                flags = struct.unpack_from('<I', cdata)[0] | (1 << 2)
                new_chunks.append((cid, struct.pack('<I', flags) + cdata[4:]))
            elif cid != b'XMP ':
                new_chunks.append((cid, cdata))
        new_chunks.append((b'XMP ', xmp_bytes))
    else:
        # Format simple (VP8 ou VP8L) — convertir en format étendu
        img = next(((cid, cd) for cid, cd in chunks if cid in (b'VP8 ', b'VP8L')), None)
        if not img:
            raise ValueError("Aucun chunk image trouvé")
        img_cid, img_data = img
        w, h = vp8_dims(img_data) if img_cid == b'VP8 ' else vp8l_dims(img_data)
        if w is None:
            raise ValueError("Impossible de lire les dimensions")
        vp8x_data = (
            struct.pack('<I', 1 << 2) +       # flags : XMP présent
            struct.pack('<I', w - 1)[:3] +    # canvas width - 1 (3 octets LE)
            struct.pack('<I', h - 1)[:3]      # canvas height - 1 (3 octets LE)
        )
        new_chunks = [(b'VP8X', vp8x_data), img, (b'XMP ', xmp_bytes)]

    Path(filepath).write_bytes(build_webp(new_chunks))

# ── Génération des métadonnées depuis le nom de fichier ─────────────────────
def meta_from_filename(fname, base_keywords, description_prefix):
    name  = fname.lower()
    label = fname.replace('.webp', '')

    # Retirer les préfixes répétitifs pour la description courte
    for prefix in [
        'Marseille-dark-massilia-plastique-pollution-projet-sentinelle-',
        'photographe-sous-marin-marseille-',
        'karim-saari-marseille-',
    ]:
        if label.startswith(prefix):
            label = label[len(prefix):]
            break

    label = label.replace('-', ' ').replace('_', ' ').strip().capitalize()

    is_underwater = any(kw in name for kw in UNDERWATER_HINTS)
    if is_underwater:
        description = (
            f"Photographie sous-marine — {label}. "
            f"Dépollution sous-marine en apnée, Calanques de Marseille — "
            f"Projet Sentinelle par Karim Saari (Dark Massilia)"
        )
        keywords = base_keywords + ["dépollution sous-marine", "freediving calanques"]
    else:
        description = f"{description_prefix}{label}. Marseille et Calanques — Karim Saari (Dark Massilia)"
        keywords = base_keywords + ["calanques de marseille", "littoral méditerranéen"]

    return label[:80], description, keywords

# ── Traitement d'un répertoire ───────────────────────────────────────────────
def process(dirpath, base_keywords, description_prefix, recursive=False):
    dirpath = Path(dirpath)
    if not dirpath.exists():
        print(f"  ⚠  Dossier introuvable : {dirpath}")
        return

    pattern = "**/*.webp" if recursive else "*.webp"
    files   = sorted(dirpath.glob(pattern))
    # Exclure les variantes srcset (_400w, _800w, _1200w) — même meta que l'original
    files   = [f for f in files if not any(s in f.name for s in ('_400w', '_800w', '_1200w'))]

    print(f"\n📁 {dirpath}  ({len(files)} fichiers)")
    ok = err = 0
    for fpath in files:
        title, desc, kws = meta_from_filename(fpath.name, base_keywords, description_prefix)
        xmp = make_xmp(title, desc, CREATOR, COPYRIGHT, kws)
        try:
            inject_xmp(fpath, xmp)
            print(f"  ✓  {fpath.name}")
            ok += 1
        except Exception as e:
            print(f"  ✗  {fpath.name} → {e}")
            err += 1
    print(f"  → {ok} OK, {err} erreurs")

# ── Main ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print("🔧  Injection XMP (sans ré-encodage)…\n")

    # Galerie photos (portfolio/Mer)
    process(
        PORTFOLIO_DIR,
        base_keywords=KEYWORDS_PORTFOLIO,
        description_prefix="Photographie Calanques de Marseille — ",
    )

    # Images missions (racine /images, seulement les Marseille-dark-massilia-*)
    # On exclut portfolio/ pour ne pas retraiter deux fois
    missions_files = sorted(
        f for f in MISSIONS_DIR.glob("Marseille-dark-massilia-*.webp")
        if not any(s in f.name for s in ('_400w', '_800w', '_1200w'))
    )
    print(f"\n📁 {MISSIONS_DIR}  ({len(missions_files)} fichiers missions)")
    ok = err = 0
    for fpath in missions_files:
        title, desc, kws = meta_from_filename(fpath.name, KEYWORDS_MISSIONS, "Mission dépollution — ")
        xmp = make_xmp(title, desc, CREATOR, COPYRIGHT, kws)
        try:
            inject_xmp(fpath, xmp)
            print(f"  ✓  {fpath.name}")
            ok += 1
        except Exception as e:
            print(f"  ✗  {fpath.name} → {e}")
            err += 1
    print(f"  → {ok} OK, {err} erreurs")

    print("\n✅  Terminé — relance npm run build:full + FTP")
